core/src/km_core_processevent_api.cpp

```diff
@@ -52,6 +52,30 @@ km_core_process_event(km_core_state *state,
   }
   km_core_status status = state->processor().process_event(state, vk, modifier_state, is_key_down, event_flags);
 
+  if (state_should_invalidate_context(state, vk, modifier_state, is_key_down, event_flags)) {
+    // clear the context and the app context
+    state->context().clear();
+    state->app_context().clear();
+
+    // we are already committed. So we need to un-commit (remove the end of the vector)
+    if (state->actions().back().type == KM_CORE_IT_END) {
+      state->actions().pop_back();
+    }
+    bool foundEmit = false;
+    km::core::action oldAction;
+    // try to put the invalidate item before the emit keystroke
```

Comment:
It'd be good to add some explanation of why we are doing this -- why do we need to add the invalidate context before emit keystroke, etc.